### ISSUE TITLE
Spurious SSDP errors on startup

### ISSUE THREAD

**Issue by terminalmage** [saltstack/salt#45056]
master.py is doing a boolean check on `self.opts['discovery]`, and assuming discovery is enabled if it resolves to `True`. But because there is a default value in `salt/config/__init__.py`, this option is therefore assumed to be enabled out-of-the-box. This doesn't seem correct, the description in the release notes mentions nothing about this feature being enabled by default, and doing so would add a hard dependency in Python 2, which I don't think was intended.

It looks like the best course of action is to set `discovery` to `False` so that it is not enabled by default.

Also, the types in [this block of code](https://github.com/saltstack/salt/blob/990c628/salt/config/__init__.py#L1165-L1182) can be removed, since `_validate_opts()` doesn't do type checking on sub-keys of a dict, and therefore these options are never going to be checked.

CC: @isbm @cachedout @thatch45

**Comment by isbm** [comment]
@terminalmage Hi!

Right. Let's see what we can do here:

- While it is a hard dependency, the library is quite old, used widely in OpenStack and so is packaged quite everywhere (pip, rpms, dpks etc). [We did that for sure](https://build.opensuse.org/package/show/openSUSE:Leap:42.3/python-trollius) long time ago, and if [it is in Debian stable](https://packages.debian.org/stretch/python-trollius), then it is everywhere. :wink:  So I am thinking why not just include it until EOL of Python2 by default? Python 3 don't need it (as `trollius` is basically a port of `asyncio`). Then we can throw it away soon anyway if we drop Python2 entirely. Additionally, enabling `discovery: True` and then seeing "ERROR: foobar is not installed" is also kind of... meh... I have no much good strategy here to offer other than indeed `Requires: python-trollius` in a package.

- The default topic. I thought I mentioned that it is by default, seems forgot it... Well, this is a question. While now it requires clean networks (or you will give the root to a malicious Master), it would be nice to have this by clean default. This is very useful if you have your (usually isolated) network of bunch of VMs of, say, cloud, and you don't want to go over each node and say which is IP/host of the Master but just batch-bootstrap them at one go (as long as the image has a Minion, started by default).

- The type-checking. Yes, it doesn't checks this _yet_. As we agreed with @cachedout a bit earlier that this is subject to change anytime soon (as it is not a major feature that even needs to be mentioned in the press release other than "now it fail for sure if you configure something completely unexpected"). So I would happily implement type check and the whole config integrity, instead of removing this.

Does it makes sense so far? Please tell me what you think.

**Comment by isbm** [comment]
Please assign me to this bug.

**Comment by terminalmage** [comment]
It's my opinion that we should be avoiding new hard dependencies unless required for core functionality. This doesn't look like core functionality to me. I'd like to hear from @cachedout and @thatch45 though.

Additionally, I just find it troubling that this was snuck in without proper discussion. Adding hard dependencies requires that we plan for it and update our own package build process. If this hadn't been caught ahead of time we'd have tons of users getting confusing errors in their logs. Issues would be opened on the bug tracker, which we would have to deal with. _This makes more work for us._

With regard to the `VALID_OPTS` issue, assuming we do update our type checking code, I don't like the fact that the discovery options are at the top level with all of the other options. This seems prone to possible collisions with other options. At minimum I would suggest changing the options I highlighted in the OP so it is clear that they are subkeys of `discovery` (for example, `discovery:mapping`, `discovery:match`, etc.).

**Comment by isbm** [comment]
> this was snuck in without proper discussion. 

We opened this PR a month ago and we were and are always looking for constructive dialogues and reviews. By no means we want to bring SaltStack more work and issues, alas we would have the same effect in our issue tracker(s).

> Adding hard dependencies requires that we plan for it and update our own package build process.

Yes, this is true and I am not even arguing this. Surely, if you have no enough time to do so, turning it off by default is not even a discussion. From our side we can simply include a dependency package in OBS and add a patch that enables this by default. Please find out what is best option for _you_ here and I will fix it ASAP. No worries.

> With regard to the `VALID_OPTS`

OK, seems I simply do not understand this part, hence this. I would not like to have any prefix, like `discovery_port` or `discovery_mapping` etc., but place them all into a namespace (sort of) under the `discovery` aggregate in this case. That said, how do I properly declare that in current implementation, so I end up with this sort of valid configuration:

Enabled:
```yaml
discovery:
    port: 5678
    mapping: 
        type: dev
        something: else
```

Disabled:
```yaml
discovery: False
```

I originally thought this:
```python
VALID_OPTS = {
    'discovery': (bool, dict),
}
```
But then `VALID_OPTS` is flat. I.e. no way to define what kind of dict structure is valid inside the `VALID_OPTS`. So I _thought_ this is like _all possible options_ are at the top level describing their _type_ (not the content), and so then later you can construct a specific structure in the defaults. But seems I was wrong in this. OK. So then how  it is usually dealt with this?

I can update this right away and problem solved. :peace_symbol:

**Comment by gtmanfred** [comment]
I also agree with not changing the default from just looking up salt.  This by default adds a 15 second delay to starting the minion.

I opened a PR on Friday to change the default to `False` here https://github.com/saltstack/salt/pull/45050

**Comment by thatch45** [comment]
Yes, we do not want to add a default hard dep to Salt for discovery. I really like having discovery, but I would much rather make it an opt-in.
I do understand the Suse argument for turning it on by default, that should be a simple addition in the Suse spec file, but I think this would be problematic to be turned on by default in a number of environments that salt is used in

**Comment by terminalmage** [comment]
Yeah, if Suse wants to turn it on by default, it's as simple as adding the dep and patching the conf file template.

**Comment by thatch45** [comment]
or deploying a config with it enabled, which would be my understanding on the best way to do it, but I might be off on Suse policy here.

**Comment by terminalmage** [comment]
Yeah... Instead of patching the template, just adding a .conf file in /etc/salt/master.d with the discovery bits would probably be easier.

**Comment by damon-atkins** [comment]
See my comments in the PR #45050 to keep everyone happy

**Comment by rallytime** [comment]
Hi @isbm - I know you have been away for the holidays, but I wanted to check in with you here to make sure this is still on your radar. We have an upcoming deadline for oxygen blockers to be complete for the first round of RC testing on Jan. 17.

**Comment by isbm** [comment]
I am back now.

@terminalmage Maybe we need to clarify this a bit: the _minion_ should not by default looking for SSDP. Which means, if we ship minion with the config `discovery: false`, then it will be OK. The SSDP could start by default on Master, since it doesn't hurt. If anyone wants to turn that off, they can explicitly set config `discovery: false`. Would that work?

**Comment by terminalmage** [comment]
You're still talking about adding a hard dependency which has already been ruled out by @thatch45. So, no, that would not work.

**Comment by terminalmage** [comment]
Whatever Suse would like to do in their packages is their business, but I think we've made it pretty clear what the project's opinions are here.

**Comment by thatch45** [comment]
Right we may have a misunderstanding here, but we will need to set discovery to False by default and not add an additional hard dep

**Comment by terminalmage** [comment]
I spoke with @isbm about his concerns with removing the default values for this feature, which I will summarize below:

1. We need to have a way for the user to have a sane default
2. Since there are multiple components nested under the `discovery`  option, there needs to be a way to allow for the user to partially configure it and still have defaults for the items they do not configure.

The solution for this would be to handle it via post-processing in `apply_master_config()` and `apply_minion_config()`. The logic would be as follows:

- If `opts['discovery']` is `False`, leave it as-is.
- If `opts['discovery']` is `True`, use the default value.
- If `opts['discovery']` is a dictionary, merge that value into the defaults and then replace `opts['discovery']` with the merged configuration.

The default values presently configured in `DEFAULT_MASTER_OPTS` and `DEFAULT_MINION_OPTS` would be replaced by `False`, and those default values would be moved into the logic in `apply_master_config()` and `apply_minion_config()`.

Finally, the `mapping`, `match`, `port`, `attempts`, and `pause` values in `VALID_OPTS` would be commented out or removed until we can come up with a more comprehensive solution for validating types in nested configurations (targeted for the Fluorine release cycle).
